orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -1179,11 +1194,65 @@ common::Status GenerateSubgraph(Graph& graph, Node* start_node) {
   return graph.Resolve();
 }
 
+Status RemoveNonDifferentiableEdgeInPartition(
+    Graph& graph,
+    const std::vector<std::string>& non_differentiable_edge_names) {
+
+  assert(!non_differentiable_edge_names.empty());
+
+  // Check a given node args vector and see if it contains non-differentiable edges. If it does, remove it.
+  auto check_node_args = [&](std::vector<NodeArg*> node_args, bool is_inputs, Node& node) {
+      auto it = node_args.begin();
+      auto begin = it;
+      auto end = node_args.end();
+      while (it != end) {
+        if (std::any_of(non_differentiable_edge_names.cbegin(), non_differentiable_edge_names.cend(), [&](std::string substr) {
+              return (*it)->Name().rfind(substr, 0) == 0;
+            })) {
+
+          node_args.erase(it);
+
+          if (is_inputs){
+            assert(!node.MutableInputArgsCount().empty() && node.MutableInputArgsCount().back() > 0);
+            node.MutableInputArgsCount().back()--;
+          }
+
+          // modify attribute to remove the erased node arg's data type
+          auto& attributes = node.GetMutableAttributes();
+          auto& element_types = attributes["element_types"];
+          if (element_types.ints_size() > 0) {
+            auto ints_copy = element_types.ints();
+            element_types.clear_ints();
+            for (auto index = 0; index < ints_copy.size(); ++index) {
+              if (index != it - begin) {
+                element_types.add_ints(ints_copy[index]);
```

Comment:
Could std::copy be used here to copy the blocks of type information before + after the one we are removing?  Actually, could we use std::erase?